app/controllers/api/v3/courses_controller.rb

```diff
@@ -3,14 +3,25 @@ module V3
     class CoursesController < API::V3::ApplicationController
       before_action :build_recruitment_cycle
       before_action :build_provider
+      before_action :build_courses
 
       def index
-        @courses = @provider.courses
-        render jsonapi: @courses, include: params[:include]
+        links = {}
+        page_index = params.dig(:page, :page)
+        per_page = params.dig(:page, :per_page)
+
+        if per_page.present?
+          @courses = @courses.page(page_index).per(per_page)
+
+          links[:next] = next_page if next_page.present?
+          links[:prev] = previous_page if previous_page.present?
+        end
```

Comment:
Reading the [docs for api-pagination](https://github.com/davidcelis/api-pagination#rails), it looks like it can handle all this header generation for us, which is kind of the point of why we're using it I think. I think all we need to do is:

```
        @courses = paginate @courses
```

Reply:
api-pagination doesn't adhere to the JSON:API standard for pagination, and so breaks Rails Find.

Reply:
Using `paginate` is a nice shortcut to do a lot of the processing that's currently being hand-done. So, if you add this to the action:

```
          @courses = paginate @courses
```

you can then add a method that `jsonapi.rb` uses to auto-generate those links:

```
      def jsonapi_pagination(collection)
        {
          first: url_for(page: 1),
          last: url_for(page: collection.total_pages),
          next: url_for(page: collection.next_page),
          prev: url_for(page: collection.prev_page),
        }
      end
```

(unfortunately it appears this isn't documented, but see the PR jsonapi-rb/jsonapi-rails#35)

With this, you get both the `Link` HTTP header:

```
Link: <http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=1&per_page=2>; rel="first", <http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=1&per_page=2>; rel="prev", <http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=15&per_page=2>; rel="last", <http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=3&per_page=2>; rel="next"
```

And the link section in JSONAPI:

```
    "links": {
        "first": "http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=1",
        "last": "http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=15",
        "next": "http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=3",
        "prev": "http://localhost:3001/api/v3/recruitment_cycles/2020/providers/1LA/courses?page=1"
    }
```

With less code. `jsonapi_pagination` can be put in a parent controller for automatic inclusion, although it'll need a guard to make sure `collection` is paginated.